packages/server/src/orchestrator/atom.ts

```diff
@@ -27,28 +27,68 @@ export class Atom {
     this.subscriptions.emit('state', this.state);
   }
 
-  // Ramda Lens types. How do they work?
-  //eslint-disable-next-line @typescript-eslint/no-explicit-any
-  view(lens: any): unknown {
-    return R.view(lens, this.get());
+  slice<T>(path: string[]): Slice<T> {
+    return new Slice(this, path);
   }
 
-  // Ramda Lens types. How do they work?
-  //eslint-disable-next-line @typescript-eslint/no-explicit-any
-  over(lens: any, fn: any) {
-    //eslint-disable-next-line @typescript-eslint/no-explicit-any
-    this.update(R.over(lens, fn) as any as (state: OrchestratorState) => OrchestratorState);
+  *next(): Operation {
+    let [state]: [OrchestratorState] = yield on(this.subscriptions, 'state');
+    return state;
   }
+}
 
-  // Ramda Lens types. How do they work?
+import { lensPath } from 'ramda';
+
+export class Slice<T> {
   //eslint-disable-next-line @typescript-eslint/no-explicit-any
-  set(lens: any, value: unknown) {
-    //eslint-disable-next-line @typescript-eslint/no-explicit-any
-    this.update(R.set(lens, value) as any as (state: OrchestratorState) => OrchestratorState);
+  lens: any;
+
+  constructor(private atom: Atom, public path: string[]) {
+    this.lens = lensPath(path);
+  }
+
+  get state() { return this.atom.get(); }
+
+  get(): T {
+    return R.view(this.lens)(this.state) as T;
+  }
+
+  set(value: T): void {
+    this.atom.update(state => {
+      return R.set(this.lens, value, state) as unknown as OrchestratorState;
+    });
+  }
+
+  over(fn: (value: T) => T): void {
+    this.atom.update(state => R.over(this.lens, fn, state) as unknown as OrchestratorState);
+  }
+
+  slice<T>(path: string[]): Slice<T> {
+    return new Slice(this.atom, this.path.concat(path));
+  }
+
+  remove(): void {
+    // If this is the root, then it cannot be removed.
+    if (this.path.length === 0) { return; }
+
+    let parentPath = this.path.slice(0, -1);
+    let parentLens = R.lensPath(parentPath);
+    let parent = R.view(parentLens, this.state);
+    if (Array.isArray(parent)) {
+      this.atom.update(state => {
+        let array = parent as unknown[];
+        return R.set(parentLens, array.filter(el => el !== this.get()), state) as unknown as OrchestratorState;
+      })
+    } else {
+      let [property] = this.path.slice(-1);
+      this.atom.update(state => {
+        return R.set(parentLens, R.dissoc(property, parent), state) as unknown as OrchestratorState;
+      })
+    }
```

Comment:
It's a bit of a smell that this ends up being so complicated.

Reply:
@cowboyd could we use microstates like api or microstates here? it would make this entire block a one liner

Reply:
We could, but microstates does not currently have any TypeScript typings. How much effort would it be to fix that?

Reply:
> It's a bit of a smell that this ends up being so complicated.

Yeah, the problem is that you need to capture the relationship between the objects when you are establishing the lens. The alternative (apart from using microstates, which abstracts over lenses, but does not currently have any type information) would be to manually walk the `path` given to `slice` and if it's a string, use `lensProp`, or if it's a number, use `lensIndex`

Reply:
> How much effort would it be to fix that?

It can be very doable if we scrap the DSL (potentially use something similar to Robbie's interactors) and drop the context funny business.